server/mem_agent/engine.py

```
@@ -332,3 +337,312 @@ def _subprocess_entry() -> None:
+
+            logger.debug(f"[SANDBOX] Return code: {result.returncode}")
+            logger.debug(f"[SANDBOX] Stderr: {result.stderr.decode()[:500]}")
+
+            if result.returncode != 0:
+                error_msg = result.stderr.decode().strip()
+                logger.error(f"[SANDBOX] Execution failed: {error_msg}")
+                return None, error_msg
+
+            # Parse output
+            try:
+                new_state, error = pickle.loads(result.stdout)
```

Ruff (0.14.13)

`pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)